Unexpected Liquid Type Mismatch

**Describe the bug**

The reproducer given below fails with a
```
**** LIQUID: UNSAFE (5 constraints checked) ************************************
Test.hs:43:1: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {v : (Test.State a) | v == s0##afzi}
    .
    is not a subtype of the required type
      VV : (Test.State a)
    .
    in the context
      s0##afzi : (Test.State a)
    Constraint id 16
   |
43 | test x = x
   | ^^^^^^^^^^^^
```

**To Reproduce**

Try to compile the following reproducer with LH:
```haskell
{-# LANGUAGE LambdaCase #-}

{-# OPTIONS_GHC -fplugin=LiquidHaskell #-}

{-@ LIQUID "--reflection" @-}

module Test where

import Prelude

{-@
data State a
  = Empty { empty :: { v : [a] | len v >= 0 } }
  | NonEmpty { nonempty :: { v : [a] | len v > 0 } }
@-}
data State a
  = Empty { empty :: [a] }
  | NonEmpty { nonempty :: [a] }

{-@ inline isEmpty @-}
isEmpty :: State a -> Bool
isEmpty = \case
  Empty{} -> True
  _       -> False

{-@
reflect charge
charge ::
  s : State a ->
  { j : Int | isEmpty s <=> j = 0 }
@-}
charge :: State a -> Int
charge = \case
  Empty{}    -> 0
  NonEmpty{} -> 1

{-@
test ::
  s0 : State a ->
  { s1 : State a | isEmpty s1 <=> charge s1 = 0 }
@-}
test :: State a -> State a
test x = x
```

This should work because the additional condition `isEmpty s1 <=> charge s1 = 0` is an invariant by construction.

**Expected behavior**

It type checks.

**Desktop (please complete the following information):**
 - OS/Distro: Nix Development Shell
 - GHC version: `9.12.2`
 - Liquid Haskell: `0.9.12.2.1`

#2691 produces the following error message, which I find preferable to the current one.

```
Test3.hs:41:1: error:
    Liquid Type Mismatch
    .
    The inferred type
      VV : {lq_tmp$x##1241 : (Test3.State a) | lq_tmp$x##1241 == x##aEx}
    .
    is not a subtype of the required type
      VV : {VV##1239 : (Test3.State a) | (if is$Test3.Empty VV##1239 then true else false) <=> Test3.charge VV##1239 == 0}
    .
    in the context
      x##aEx : (Test3.State a)
    Constraint id 16
   |
41 | test x = x
   | ^^^^^^^^^^
```

I'd close this issue after merging this, as I don't think we have a well defined way to address the behavior differences between measure/inline/reflect. Could be worth writing a blog post and linking that from the documentation, maybe.

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/UX/Errors.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/UX/Errors.hs
@@ -122,11 +122,7 @@ expandVarDefs      = go mempty
 
 isInline :: (a, SpecType) -> Either (a, F.Expr) (a, SpecType)
 isInline (x, t) =
-    either (Left . (x,)) (Right . (x,)) (isInline' t')
-  where
-    -- tidyInternalRefas could drop some conjuncts which affects whether
-    -- bindings are eliminated in isInline'
-    t' = tidyInternalRefas t
+    either (Left . (x,)) (Right . (x,)) (isInline' t)
 
 isInline' :: SpecType -> Either F.Expr SpecType
 isInline' t = case ro of

--- a/liquidhaskell-boot/src/Language/Haskell/Liquid/UX/Tidy.hs
+++ b/liquidhaskell-boot/src/Language/Haskell/Liquid/UX/Tidy.hs
@@ -14,7 +14,6 @@ module Language.Haskell.Liquid.UX.Tidy (
 
     -- * Tidying functions
     tidySpecType
-  , tidyInternalRefas
   , tidySymbol
 
     -- * Panic and Exit
@@ -85,7 +84,6 @@ tidySpecType k
   . tidyValueVars
   . tidyDSymbols
   . tidySymbols k
-  . tidyInternalRefas
   . tidyLocalRefas k
   . tidyFunBinds
   . tidyTyVars
@@ -127,16 +125,6 @@ tidyEqual = mapReft txReft
   where
     txReft u                      = u { ur_reft = mapPredReft dropInternals $ ur_reft u }
     dropInternals                 = pAnd . L.nub . conjuncts
-
--- | Drop conjuncts that contain data constructor testing or
--- selector functions.
-tidyInternalRefas   :: SpecType -> SpecType
-tidyInternalRefas = mapReft txReft
-  where
-    txReft u                      = u { ur_reft = mapPredReft dropInternals $ ur_reft u }
-    dropInternals                 = pAnd . filter (not . any isIntern . syms) . conjuncts
-    isIntern x                    = "is$" `isPrefixOfSym` x || "$select" `isSuffixOfSym` x
-
 
 tidyDSymbols :: SpecType -> SpecType
 tidyDSymbols t = mapBind tx $ substa tx t

--- a/tests/errors/T2650.hs
+++ b/tests/errors/T2650.hs
@@ -1,0 +1,42 @@
+{-# LANGUAGE LambdaCase #-}
+{-# OPTIONS_GHC -fplugin=LiquidHaskell #-}
+
+{-@ LIQUID "--reflection" @-}
+
+-- | Tests that error messages show predicates that mix user-defined functions
+-- with internal data constructor tests (is$) from inlined definitions.
+-- See https://github.com/ucsd-progsys/liquidhaskell/issues/2650
+--
+-- Before the fix, the error would show just:
+--   VV : T2650.State
+-- (dropping the predicate entirely because it contained is$T2650.Empty).
+--
+-- After the fix, the required type shows the full predicate:
+--   VV : {... | (if is$T2650.Empty VV then true else false) <=> T2650.isGood VV}
+
+{-@ LIQUID "--expect-error-containing=T2650.isGood" @-}
+
+module T2650 where
+
+import Prelude
+
+data State = Empty | NonEmpty
+
+{-@ inline isEmpty @-}
+isEmpty :: State -> Bool
+isEmpty = \case
+  Empty    -> True
+  NonEmpty -> False
+
+{-@ reflect isGood @-}
+isGood :: State -> Bool
+isGood Empty    = True
+isGood NonEmpty = False
+
+{-@
+test ::
+  s0 : State ->
+  { s1 : State | isEmpty s1 <=> isGood s1 }
+@-}
+test :: State -> State
+test x = x

--- a/tests/tests.cabal
+++ b/tests/tests.cabal
@@ -668,6 +668,7 @@ executable errors
                     , T1498A
                     , T1498
                     , T2346
+                    , T2650
                     , T773
                     , T774
                     , TerminationExprNum

